tests/core/test_typecheck.py

```diff
@@ -0,0 +1,436 @@
+
+        assert result2.sum() == torch.tensor(10 * 5)
+        assert hasattr(result, 'neural_type') is False
+
+    @pytest.mark.unit
+    def test_input_output_neural_types(self):
+        class NodeA(Typing):
+            @property
+            def input_types(self):
+                return {"x": NeuralType(('B',), ElementType())}
+
+            @property
+            def output_types(self):
+                return {"y": NeuralType(('B', 'D'), LogitsType())}
+
+            @typecheck()
+            def __call__(self, x):
+                y = torch.randn(x.shape[0], 4)
+                return y
+
+        class NodeB(Typing):
+            @property
+            def input_types(self):
+                return {"w": NeuralType(('B', 'D'), LogitsType())}
+
+            @property
+            def output_types(self):
+                return {"u": NeuralType(('B',), LabelsType())}
+
+            @typecheck()
+            def __call__(self, w):
+                _, u = w.max(-1)
+                return u
+
+        nodeA = NodeA()
+        nodeB = NodeB()
+
+        outA = nodeA(x=torch.zeros(10))
+        outB = nodeB(w=outA)
+
+        assert outB.shape == torch.Size([10])
+        assert outB.neural_type.compare(NeuralType(('B',), LabelsType())) == NeuralTypeComparisonResult.SAME
+
+    @pytest.mark.unit
+    def test_nested_input_output_neural_types(self):
+        class NestedNodeA(Typing):
+            @property
+            def input_types(self):
+                return {"x": NeuralType(('B',), ElementType())}
+
+            @property
+            def output_types(self):
+                return {
+                    "y0": NeuralType(('B', 'D'), LogitsType()),
+                    "y1": NeuralType(('B', 'D'), LogitsType()),
+                }
+
+            @typecheck()
+            def __call__(self, x):
+                # input x = [[x1, x2], [x3]]
+                x0 = x[0][0]
+                y = torch.randn(x0.shape[0], 4)
+                return [[y, y], [y]]
+
+        # Non-homogeneous output types
+        class NestedNodeB(Typing):
+            @property
+            def input_types(self):
+                return {"w": NeuralType(('B', 'D'), LogitsType())}
+
+            @property
+            def output_types(self):
+                return {
+                    "u0": NeuralType(('B',), LogprobsType()),  # check non homogeneous type
+                    "u1": NeuralType(('B',), LabelsType()),
+                }
+
+            @typecheck()
+            def __call__(self, w):
+                # input x = [[x1, x2], [x3]]
+                _, u00 = w[0][0].max(-1)
+                _, u01 = w[0][1].max(-1)
+                _, u10 = w[1][0].max(-1)
+                return [[u00, u01], [u10]]
+
+        nodeA = NestedNodeA()
+        nodeB = NestedNodeB()
+
+        input_nest = [[torch.zeros(10), torch.zeros(10)], [torch.zeros(10)]]
+        outA = nodeA(x=input_nest)
```

Comment:
@titu1994 @okuchaiev I think this call should return an error: the input matching the definition: `NeuralType(('B',), ElementType()` should be `torch.zeros(10)` not `list of list of torch.zeros(10)`.

`List of list of torch.zeros(10)` should have this definition (incorrect syntax, just to show what I have in mind):
```
NeuralType(
  (Batch, is_list=True),
  (Time, is_list=True),
  (Depth, len=10),
   ElementType()
 )
```

This PR changes how type checking handles "dynamic axes" (i.e. axes that are lists). It seems that “dynamic” axes are skipped when type checking…? I really do not understand what is the logic of type checking now...